File: rcf/imas-backend/src/main/java/com/rcf/imas/modules/tracking/web/TrackingController.java
package com.rcf.imas.modules.tracking.web;

import com.rcf.imas.modules.tracking.persistence.TrackingReadRepository;
import com.rcf.imas.platform.error.ApiException;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/tracking")
@PreAuthorize("hasRole('ADMIN')")   // audit CRITICAL: Node applies zero auth middleware to this mount (Firm Decision 1)
class TrackingController {

    private final TrackingReadRepository reads;
    private final String fileStoragePath;

    TrackingController(TrackingReadRepository reads, @Value("${imas.file-storage-path}") String fileStoragePath) {
        this.reads = reads;
        this.fileStoragePath = fileStoragePath;
    }

    @GetMapping("/interviewers")
    public List<Map<String, Object>> interviewers() {
        try {
            return reads.allInterviewers();
        } catch (Exception e) {
            throw ApiException.error(500, "Could not fetch interviewers.");
        }
    }

    @GetMapping("/students/interviewer/{interviewerId}")
    public Map<String, Object> studentsByInterviewer(@PathVariable String interviewerId,
                                                       @RequestParam(defaultValue = "1") int page,
                                                       @RequestParam(name = "nmms_year", defaultValue = "2025") String nmmsYear) {
        if (!interviewerId.matches("\\d+")) {
            throw ApiException.error(400, "Invalid Interviewer ID provided.");
        }
        try {
            return reads.studentsByInterviewer(interviewerId, page, nmmsYear);
        } catch (Exception e) {
            throw ApiException.error(500, "Could not fetch students assigned to interviewer.");
        }
    }

    @GetMapping("/students/{applicantId}/details")
    public List<Map<String, Object>> studentDetails(@PathVariable String applicantId,
                                                      @RequestParam(name = "nmms_year", defaultValue = "2025") String nmmsYear,
                                                      @RequestParam(required = false) String filtered) {
        // filtered=true is INERT (quirk 4b) -- both branches call the identical repository method.
        if (!applicantId.matches("\\d+")) {
            throw ApiException.error(400, "Invalid Applicant ID.");
        }
        try {
            List<Map<String, Object>> rows = reads.studentDetailForFilter(applicantId, nmmsYear);
            if (rows.isEmpty()) {
                throw ApiException.error(404, "Student or interview data not found.");
            }
            return rows;
        } catch (ApiException e) {
            throw e;
        } catch (Exception e) {
            throw ApiException.error(500, "Could not fetch student interview details.");
        }
    }

    @GetMapping("/students/{applicantId}/interviews/all")
    public List<Map<String, Object>> allInterviewRounds(@PathVariable String applicantId,
                                                          @RequestParam(name = "nmms_year", defaultValue = "2025") String nmmsYear) {
        if (!applicantId.matches("\\d+")) {
            throw ApiException.error(400, "Invalid Applicant ID.");
        }
        try {
            return reads.allInterviewRounds(applicantId, nmmsYear);
        } catch (Exception e) {
            throw ApiException.error(500, "Could not fetch all interview rounds.");
        }
    }

    @GetMapping("/students/{applicantId}/home/all")
    public List<Map<String, Object>> allHomeVerificationRounds(@PathVariable String applicantId) {
        if (!applicantId.matches("\\d+")) {
            throw ApiException.error(400, "Invalid Applicant ID.");
        }
        try {
            return reads.allHomeVerificationRounds(applicantId);
        } catch (Exception e) {
            throw ApiException.error(500, "Could not fetch home verification records.");
        }
    }

    @GetMapping("/students")
    public Map<String, Object> students(@RequestParam(defaultValue = "1") int page,
                                         @RequestParam(name = "statuses", required = false) List<String> statuses,
                                         @RequestParam(required = false) List<String> results,
                                         @RequestParam(name = "nmms_year", defaultValue = "2025") String nmmsYear) {
        // Firm Decision 7: the 2025 literal is kept (matches Node's req.query.nmms_year || 2025, ground
        // truth §7 quirk 17) rather than moved to config -- faithful-parity phase; flagged in Deferred.
        try {
            return reads.studentsWithLatestStatus(page, statuses, results, nmmsYear);
        } catch (Exception e) {
            throw ApiException.error(500, "Could not fetch student tracking data.");
        }
    }

    /**
     * downloadDocument -- file-serving redirect (Firm Decision 6). This is the ONE endpoint in the whole
     * module with plain-text error bodies (matching Node's res.send(...) calls), not the module's usual
     * {error:...} JSON envelope -- handled directly here rather than via ApiException/GlobalExceptionHandler.
     */
    @GetMapping("/document/{applicantId}/{cohortId}")
    public ResponseEntity<?> downloadDocument(@PathVariable String applicantId, @PathVariable String cohortId,
                                               @RequestParam(required = false) String type) {
        if (!applicantId.matches("\\d+") || cohortId.isBlank()
                || !("interview".equals(type) || "home".equals(type))) {
            return ResponseEntity.badRequest().contentType(MediaType.TEXT_PLAIN)
                    .body("Invalid parameters.");
        }
        try {
            Map<String, Object> meta = "interview".equals(type)
                    ? reads.interviewDocument(applicantId)
                    : reads.homeVerificationDocument(applicantId);
            if (meta == null || meta.get("doc_name") == null) {
                return ResponseEntity.status(404).contentType(MediaType.TEXT_PLAIN)
                        .body("Document metadata not found.");
            }
            String rawDocName = String.valueOf(meta.get("doc_name"));
            // basic traversal guard: strip to the last path segment on either separator (Firm Decision 6)
            String[] parts = rawDocName.split("[\\\\/]");
            String cleanDocName = parts.length == 0 ? rawDocName : parts[parts.length - 1];

            String folder = "interview".equals(type) ? "Interview-data" : "home-verification-data";
            Path onDisk = Path.of(fileStoragePath, folder, cohortId, cleanDocName);
            if (!Files.exists(onDisk)) {
                return ResponseEntity.status(404).contentType(MediaType.TEXT_PLAIN)
                        .body("File not found on storage.");
            }
            String location = "/Data/" + folder + "/" + cohortId + "/" + cleanDocName;
            return ResponseEntity.status(302).header(HttpHeaders.LOCATION, location).build();
        } catch (Exception e) {
            return ResponseEntity.status(500).contentType(MediaType.TEXT_PLAIN)
                    .body("Server Error.");
        }
    }
}

File: server/controllers/trackingController.js
/** * @fileoverview Express controllers for handling API requests for student and interview tracking.
 */ 
const path = require('path'); 
const fs = require('fs'); 
const trackingModel = require('../models/trackingModel');

const PC_STORAGE_ROOT = process.env.FILE_STORAGE_PATH; 
const INTERVIEW_DOC_DIR = 'Interview-data';
const HOME_VERIFICATION_DOC_DIR = 'home-verification-data';

// --- Utility Function ---
const constructFilePath = (docTypeFolder, doc_name, cohortFolder) => {
    return path.join(
        PC_STORAGE_ROOT,
        docTypeFolder, 
        String(cohortFolder), 
        doc_name
    );
};

const trackingController = {
    
    async getAllInterviewers(req, res) {
        try {
            const interviewers = await trackingModel.getAllInterviewers();
            res.json(interviewers);
        } catch (error) {
            console.error("Error in getAllInterviewers:", error.message);
            res.status(500).json({ error: "Could not fetch interviewers." });
        }
    },

    async getStudents(req, res) {
        const page = parseInt(req.query.page) || 1;
        const limit = 10;
        // Extract NMMS Year from query
        const nmmsYear = req.query.nmms_year || 2025; 
        
        const statuses = req.query.statuses ? req.query.statuses.split(',').filter(s => s.trim() !== '') : [];
        const resultsRaw = req.query.results ? req.query.results.split(',').filter(s => s.trim() !== '') : [];

        const homeVerificationSelected = resultsRaw.includes('HOME VERIFICATION REQUIRED');
        const results = resultsRaw.filter(r => r !== 'HOME VERIFICATION REQUIRED');

        try {
            const data = await trackingModel.getStudentsWithLatestStatus(
                page, 
                limit, 
                statuses, 
                results, 
                homeVerificationSelected,
                nmmsYear // Pass year to model
            );

            res.json({
                students: data.students,
                currentPage: page,
                totalPages: data.totalPages,
                totalStudents: data.totalCount
            });

        } catch (error) {
            console.error("Error fetching students:", error.message);
            res.status(500).json({ error: "Could not fetch student tracking data." });
        }
    },

    async getStudentsByInterviewer(req, res) {
        const interviewerId = parseInt(req.params.interviewerId);
        const page = parseInt(req.query.page) || 1;
        const limit = 10;
        const nmmsYear = req.query.nmms_year || 2025; // Extract Year

        if (isNaN(interviewerId)) {
            return res.status(400).json({ error: "Invalid Interviewer ID provided." });
        }

        try {
            const data = await trackingModel.getStudentsByInterviewer(interviewerId, page, limit, nmmsYear);

            res.json({
                students: data.students,
                currentPage: page,
                totalPages: data.totalPages,
                totalStudents: data.totalCount
            });

        } catch (error) {
            console.error(`Error fetching students for interviewer ${interviewerId}:`, error.message);
            res.status(500).json({ error: "Could not fetch students assigned to interviewer." });
        }
    },

    async getStudentDetails(req, res) {
        const applicantId = parseInt(req.params.applicantId);
        const nmmsYear = req.query.nmms_year || 2025; // Extract Year
        const isFilteredView = req.query.filtered === 'true'; 

        if (isNaN(applicantId)) {
            return res.status(400).json({ error: "Invalid Applicant ID." });
        }

        try {
            let rounds;
            if (isFilteredView) {
                rounds = await trackingModel.getStudentdetailforFilter(applicantId, nmmsYear);
            } else { 
                rounds = await trackingModel.getStudentdetailforFilter(applicantId, nmmsYear);
            }
            
            if (rounds.length === 0) {
                return res.status(404).json({ error: "Student or interview data not found." });
            }
            res.json(rounds); 

        } catch (error) {
            console.error("Error fetching student details:", error.message);
            res.status(500).json({ error: "Could not fetch student interview details." });
        }
    },

    async getAllInterviewRounds(req, res) {
        const applicantId = parseInt(req.params.applicantId);
        const nmmsYear = req.query.nmms_year || 2025; // Extract Year

        if (isNaN(applicantId)) {
            return res.status(400).json({ error: "Invalid Applicant ID." });
        }
        try {
            const rounds = await trackingModel.getAllInterviewRounds(applicantId, nmmsYear);
            res.json(rounds);
        } catch (error) {
            console.error(`Error fetching all interviews for ${applicantId}:`, error.message);
            res.status(500).json({ error: "Could not fetch all interview rounds." });
        }
    },

    async getAllHomeVerificationRounds(req, res) {
        const applicantId = parseInt(req.params.applicantId);
        // Note: home_verification table usually doesn't need nmms_year if applicant_id is unique, 
        // but it's here if your table logic requires it.
        if (isNaN(applicantId)) {
            return res.status(400).json({ error: "Invalid Applicant ID." });
        }
        try {
            const verifications = await trackingModel.getAllHomeVerificationRounds(applicantId);
            res.json(verifications); 
        } catch (error) {
            console.error(`Error fetching all home verifications for ${applicantId}:`, error.message);
            res.status(500).json({ error: "Could not fetch home verification records." });
        }
    },

    async downloadDocument(req, res) {
        const applicantId = parseInt(req.params.applicantId);
        let cohortFolder = req.params.cohortId; 
        const docType = req.query.type; 

        if (isNaN(applicantId) || !cohortFolder || !['interview', 'home'].includes(docType)) {
            return res.status(400).send('Invalid parameters.');
        }

        try {
            let docInfo;
            let dataFolder;

            if (docType === 'home') {
                docInfo = await trackingModel.getHomeVerificationDocument(applicantId);
                dataFolder = HOME_VERIFICATION_DOC_DIR;
            } else { 
                docInfo = await trackingModel.getInterviewDocument(applicantId);
                dataFolder = INTERVIEW_DOC_DIR;
            }

            if (!docInfo || !docInfo.doc_name) {
                return res.status(404).send(`Document metadata not found.`);
            }

            let { doc_name } = docInfo;
            const pathSegments = doc_name.split(/\\|\//).filter(s => s); 
            if (pathSegments.length > 0) doc_name = pathSegments.pop();

            const absoluteFilePath = constructFilePath(dataFolder, doc_name, cohortFolder);
            
            if (!fs.existsSync(absoluteFilePath)) {
                return res.status(404).send(`File not found on storage.`);
            }
            
            const publicUrlPath = path.posix.join('/Data', dataFolder, String(cohortFolder), doc_name);
            return res.redirect(publicUrlPath);

        } catch (err) {
            console.error('Error in downloadDocument:', err);
            res.status(500).send('Server Error.');
        }
    }
};

module.exports = trackingController;